### ISSUE TITLE
Reduce palette size of themes

### ISSUE THREAD

**Issue by clarfonthey** [go-gitea/gitea#19353]
### Feature Description

Right now, the [`_base.less` file](https://github.com/go-gitea/gitea/blob/main/web_src/less/_base.less) defines all of the colours for the base theme (which is overridden for the dark theme), and the current palette size is… 121 colours, with 105 unique colours; some are duplicated but not referencing each other.

Honestly, I'd say it's a big stretch that this should be larger than 20, but the exact number isn't nearly as important. We definitely have too many.

Auditing things like colour contrast is basically not possible with this many combinations, and if someone wants to make simple changes (e.g. changing just the accent colour), this isn't really feasible, since there are so many different shades of each colour.

I think it's extremely worthwhile seeing what can be trimmed from the palette by either using a similar shade or opacity to blend with the background (in cases like the heatmap).

**Comment by clarfonthey** [comment]
I agree with needlessly eliminating variables but disagree completely with the required number of shades. At a minimum, since the stylesheet is using LESS, it should be computing required shades if necessary rather than hard-coding them, since it makes it difficult to discern what would need to be changed in case someone wanted to modify the colours.

**Comment by silverwind** [comment]
> since the stylesheet is using LESS, it should be computing required shades if necessary rather than hard-coding them

Computing colors in LESS is something I certainly want to avoid because those variables can not be interacted with at runtime (we already do that as part of the monaco themeing) and it makes 3rd party theming a lot more complex as they'd need to depend on our LESS sources. Imho, variables inside the styling must strictly stay custom properties.

CSS custom properties integrate much better than preprocessor variables. It's a shame that standard CSS still does not offer a way to modify colors at runtime, but there have been some recent proposals for just that in W3C, so maybe we'll get something in 5-10 years.

**Comment by clarfonthey** [comment]
I think you're misunderstanding what I'm saying-- take the existing CSS variable definitions and replace the actual colours with versions computed using LESS functions. If you reference the compiled CSS, you'll still have static colours, but if you reference the LESS, you'll know where the colours came from.

**Comment by silverwind** [comment]
We've migrated from Less to CSS now. Once [color-mix](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix) is well supported in browsers, it can be used to reduce the number of colors, but I take it will be a year or two until that point.

**Comment by clarfonthey** [comment]
This still doesn't respond at all to the argument that the stylesheet could due with substantially fewer shades than it has.

If we're talking about super-granular case like the heatmap in user profiles, this could easily be accomplished today with `opacity` instead of explicitly setting the background to a mixed colour. In all other cases, I still strongly disagree that a large number of shades are needed for standard UI elements; even a user with very good vision would be hard-pressed to identify ten different shades of a colour on all the standard UI elements, and much more than four is really pushing it.

I mentioned using LESS as a starting point for this issue, not the ending point.

**Comment by wxiaoguang** [comment]
I agree that there are too many colors. Is there a feasible plan about how to cleaning them up?

**Comment by wxiaoguang** [comment]
And actually, I have slight objection for such UI ( #22864) , I am not a fan of making UI too colorful .....

It's better to have some widely-accepted UI design patterns.

![image](https://user-images.githubusercontent.com/2114189/235829284-49743b86-1d4f-4556-94ef-892a0358c7f0.png)

**Comment by clarfonthey** [comment]
Another big reason against those kinds of changes is that the contrast tends to be rather poor. For example, the contrast between the red and green in the bar next to "README.md" there is effectively zero if you're red-green colourblind, and the contrast between the yellow text and the white background is also very poor.

This is also the main reason why I disagree with so many shades of colours. For example, the background colour of the header for the file element is #f7f7f7, and the colour of the main background is #ffffff. This is such a low contrast that it's effectively visible *even* for someone with a good monitor and good vision. This colour should be reduced to a darker shade or replaced with the original colour.

**Comment by wxiaoguang** [comment]
Luckily, the green-red gauge only provides some minor information, it only makes the UI looks slightly beautiful.

The real problem for me here is the orange color, I am the poor man who doesn't have a good monitor / good vision .... it's difficult for me to read how many files get changed at first glance .... 🤔

**Comment by wxiaoguang** [comment]
@clarfonthey Now there are new requirements for colors (focus/active), do you have more ideas about it? #24507

**Comment by silverwind** [comment]
These are indiscussable I'd say. Fomantic UI has 4 different colors for button states, if we don't overwrite all, inconsistent colors will show up on buttons.

**Comment by clarfonthey** [comment]
The discussion is about whether different states should be different colours, not whether we should set the colours in the CSS. I think we're both with you on making sure they actually follow the theme correctly, but the question is what the theme should be in the first place.

**Comment by silverwind** [comment]
Interesting tidbit, number of color variables comparing GitHub and Gitea. GitHub has major spaghetti going on in their CSS variables, we are not nearly at that point, thankfully.

### GitHub

```js
Object.entries(getComputedStyle(document.documentElement)).filter(e => e[1].startsWith("--") && e[1].includes("color")).length
489
```

### Gitea
```js
> Object.entries(getComputedStyle(document.documentElement)).filter(e => e[1].startsWith("--") && e[1].includes("color")).length
224
```

**Comment by silverwind** [comment]
Will be fixed with https://github.com/go-gitea/gitea/issues/19353 which will give us total color freedom.

**Comment by silverwind** [comment]
Closing in favor of https://github.com/go-gitea/gitea/issues/30160.
